singer_sdk/sinks/sql.py

```diff
@@ -82,11 +82,31 @@ def schema_name(self) -> Optional[str]:
         Returns:
             The target schema name.
         """
+        # Get the current SQL Dialect being used
+        target_sqla_dialect = self.connection.engine.dialect.name
+        # Look for a default_target_scheme in the configuraion fle
+        default_target_schema = self.config.get("default_target_schema", None)
         parts = self.stream_name.split("-")
+
+        # 1) When default_target_scheme is in the configuration use it
+        # 2) if the streams are in <schema>-<table> format use the
+        #    stream <schema>
+        # 3) Return None if you don't find anything
+        if default_target_schema:
+            return default_target_schema
+
         if len(parts) in {2, 3}:
             # Stream name is a two-part or three-part identifier.
             # Use the second-to-last part as the schema name.
-            return self.conform_name(parts[-2], "schema")
+            stream_schema = self.conform_name(parts[-2], "schema")
+
+            # MS SQL Server has a public database role so the name is reserved
+            # and it can not be created as a schema.  To avoid this common error
+            # we convert "public" to "dbo" if the target dialet is mssql
+            if target_sqla_dialect == "mssql" and stream_schema == "public":
+                return "dbo"
```

Comment:
I'm not sure if the SDK should be aware of this mssql restriction other than providing a way to override/extend functionality to handle it. Without this check, an example implementation in a target-mssql could be:

```python
class MSSQLSink(SQLSink):
    def schema_name(self):
        stream_schema = super().schema_name()
        if self.connector._dialect == "mssql" and stream_schema == "public":
            return "dbo"
        else:
     	    return stream_schema
```

Reply:
Good call.  I will get this moved into the `MSSQLSink` and out of `SQLSink`.  Thanks for the code 🙏

Reply:
@BuzzCutNorman I removed the checking for `stream_schema == "public"` for mssql from the SDK and got it added into my target-mssql.